src/main/kotlin/uk/gov/communities/prsdb/webapp/services/EpcLookupService.kt

```
@@ -0,0 +1,69 @@
+package uk.gov.communities.prsdb.webapp.services
+
+import org.json.JSONObject
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Service
+import org.springframework.web.server.ResponseStatusException
+import uk.gov.communities.prsdb.webapp.clients.EpcRegisterClient
+import uk.gov.communities.prsdb.webapp.helpers.DateTimeHelper
+import uk.gov.communities.prsdb.webapp.models.dataModels.AddressDataModel
+import uk.gov.communities.prsdb.webapp.models.dataModels.EpcDataModel
+
+@Service
+class EpcLookupService(
+    private val epcRegisterClient: EpcRegisterClient,
+) {
+    fun getEpcByCertificateNumber(certificateNumber: String): EpcDataModel? {
+        val formattedCertificateNumber = EpcDataModel.formatCertificateNumber(certificateNumber)
+        val response = epcRegisterClient.getByRrn(formattedCertificateNumber)
+        val jsonResponse = JSONObject(response)
+
+        if (jsonResponse.has("errors")) {
+            val errorCode = getErrorCode(jsonResponse)
+            if (errorCode == "NOT_FOUND") {
+                return null
+            }
+            if (errorCode == "INVALID_REQUEST" || errorCode == "BAD_REQUEST") {
+                throw ResponseStatusException(HttpStatus.BAD_REQUEST, getErrorMessage(jsonResponse))
+            }
+        }
```

Rather than repeating this in the `EpcRegisterClient`, could the client handle all the error throwing and returning null for 404s? If you go with this the `getErrorCode` and `getErrorMessage` could be private `EpcRegisterClient` methods.

---

I think I prefer to leave it as the client returning the response and the service layer deciding what to do with it (e.g. there may be times when we want to throw an error for a NOT_FOUND response rather than returning null)